ambry-store/src/main/java/com/github/ambry/store/Log.java

```diff
@@ -439,14 +451,75 @@ private void free(LogSegment logSegment, boolean isSwapSegment) throws StoreExce
   private void rollOverIfRequired(long writeSize) throws StoreException {
     if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
       ensureCapacity(writeSize);
-      // this cannot be null since capacity has either been ensured or has thrown.
-      LogSegment nextActiveSegment = segmentsByName.higherEntry(activeSegment.getName()).getValue();
-      logger.info("Rolling over writes to {} from {} on write of data of size {}. End offset was {} and capacity is {}",
-          nextActiveSegment.getName(), activeSegment.getName(), writeSize, activeSegment.getEndOffset(),
-          activeSegment.getCapacityInBytes());
-      activeSegment.dropBufferForAppend();
-      nextActiveSegment.initBufferForAppend();
-      activeSegment = nextActiveSegment;
+      setActiveSegment();
+    }
+  }
+
+  /**
+   * Closes the last log segment periodically if replica is in sealed status.
+   * @param journal The {@link Journal} needs to be clean up after log segment closed.
+   * @throws StoreException if any store exception occurred as part of ensuring capacity.
+   */
+  void autoCloseLastLogSegmentIfQualified(Journal journal) throws StoreException {
+    if (compactionPolicySwitchInfoCounterValueReached()) {
+      //ensure the capacity to open the new log segment and allocate new log segment.
+      //if not able to close the last log segment, continue running the compaction.
+      try {
+        ensureCapacity(0);
+        setActiveSegment();
+        //refresh journal.
+        long startTime = SystemTime.getInstance().milliseconds();
+        if (config.storeSynchronizerInsideWhileLoopEnabled) {
+          journal.cleanUpJournalWithSynchronizeInside(activeSegment.getName());
+        } else {
+          journal.cleanUpJournal(activeSegment.getName());
+        }
+        logger.debug("Time to clean up journal size for store : {} in dataDir: {} is {} ms", storeId, dataDir,
+            SystemTime.getInstance().milliseconds() - startTime);
+      } catch (IllegalStateException e) {
+        //if there is no more capacity left in data dir to create new log segment, just bypass this logic.
+        //no-op
+      }
+    }
+  }
+
+  /**
+   * Set the last log segment as active segment and init buffer for append.
+   * @throws StoreException
+   */
+  private void setActiveSegment() throws StoreException {
+    // this cannot be null since capacity has either been ensured or has thrown.
+    LogSegment nextActiveSegment = segmentsByName.higherEntry(activeSegment.getName()).getValue();
+    logger.info("Auto close last log segment: {} and create new active segment : {}", activeSegment.getName(),
```

Comment:
This method could be called by a regular `rollOverIfRequired()` as well, the info message here will be confusing in that case.

Reply:
Thx for catching that. Added both logs.